.github/workflows/ci.yml

```
@@ -211,3 +211,42 @@ jobs:
         FULL_FEATURE=$(cargo metadata --format-version=1 --no-deps | jq -r '.packages[] | select(.name == "libp2p") | .features["full"] | sort | join(" ")')
         
         test "$ALL_FEATURES = $FULL_FEATURE"
+  gather_workspace_members:
+    runs-on: ubuntu-latest
+    outputs:
+      members: ${{ steps.cargo-metadata.outputs.members }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - id: cargo-metadata
+        run: |
+          WORKSPACE_MEMBERS=$(cargo metadata --format-version=1 --no-deps | jq -c '.packages |.[] | .name'| grep 'libp2p-' | jq -s | jq -c '.')
```

Not sure how familiar you are with `jq` but it does have a filter mechanism (called `select`) as well, so we don't need to break up the expression only pipe it to `jq` again directly after :)

```suggestion
          WORKSPACE_MEMBERS=$(cargo metadata --format-version=1 --no-deps | jq -c '.packages |.[] | select(.name | startsWith("libp2p-")')
```

---

Really, the even better way would be to select them based on the `publish` value under `package` in `Cargo.toml`.

We want to check all semver compliance for all packages that we publish to crates.io. That f.e. includes `multistream-select` which isn't currently included in this.

I think the above should give you the necessary tools to get that done! Happy jq hacking 😁

---

Thank you for being so detailed in your suggestions :) I'm not so familiar with `jq` so this is a very enlightening experience! Very cool. 

I don't think it would be wise to use the publish value considering they are all `null`-

```bash
cargo metadata --format-version=1 --no-deps | jq -c '.packages | .[] | .publish'
``` 
shows all `null` values -  which makes sense they were never given a value in the manifests, and the [publish field is actually used to prevent it  from being published to a package registry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field).  

We could actually run `semver-checks` on all the packages but it will throw an error for new packages since there would be no available baseline versions f.e. with `keygen` or `quickcheck-ext`. Perhaps we should run it on all packages greater than `0.1.0` ?

---

The reason why I proposed using `publish` is because we only care about semver in packages we publush :)

If you set it to `false` in keygen, does it show up then? It is a shame cargo doesn't autofill defaults for the  manifests. The default for `publish` is `true`. I'd be fine with having that logic in here and selecting all crates that have either `null` or `true` for their `publish` value.

---

I agree it is a shame, I opened [this issue](https://github.com/oli-obk/cargo_metadata/issues/207) to address it. In the interim we can select on a `null` value.

---

> We could actually run semver-checks on all the packages but it will throw an error for new packages since there would be no available baseline versions f.e. with keygen or quickcheck-ext. Perhaps we should run it on all packages greater than 0.1.0 ?

Also how do you think this is a reasonable suggestion?

---

> > We could actually run semver-checks on all the packages but it will throw an error for new packages since there would be no available baseline versions f.e. with keygen or quickcheck-ext. Perhaps we should run it on all packages greater than 0.1.0 ?
> 
> Also how do you think this is a reasonable suggestion?

I don't think it is a big problem that it throws an error on new packages. That is a temporary state when adding a new crate so it shouldn't happen all too often. @mxinden Do you think it is acceptable to have a failing job between merging the PR for a new crate and releasing it?

Selecting based on the version is a bit of a hack because it is not directly tied to the issue. Someone could add a package with version `0.0.1` if they want so if we were to go by version, we'd have to select based on `> 0.0.0` really (and have all sample / dev crates use `0.0.0`). Using `publish` is a more direct link to what we want to enforce: Not violating semver for packages we publish to `crates.io`.

---

Let me clarify, I was suggesting that if there is no `baseline version` for a crate to us to compare to, then we would avoid doing a `semver check` - _since there would be no version for us to check against._ 

This is not to suggest that we should select by `version` as an alternative to `publish` , but rather that we do this check per package that we publish.

Otherwise this would continuously fail on packages like `keygen` for example (which is over 3 years old) as it has only one version.

---

Ah right, I misunderstood you!

So we would build the matrix from all publishable crates but then check if we actually have a baseline version? How would we achieve that in particular? If the crate is published as `0.1.0` at the moment, we still want to check.

To me, this sounds like a missing feature in `cargo semver-checks`: We need a flag to have it not fail with missing baseline version to make it more CI friendly.

---

Good point, upon reconsideration I don't think there is a version detection that could indicate that there wasn't a semver violation in a package

I raised an issue on the repo https://github.com/obi1kenobi/cargo-semver-check/issues/146 

In the interim we could ignore the failures that are related to `No available baseline versions` or ignore those packages manually. What do you think @thomaseizinger ?

---

Given that they are all separate jobs, I think we can just have those fail for the time period between merging a new crate and releasing it :)